---
title: "RNA-seq report"
output:
  pdf_document:
    fig_caption: yes
  html_document: default
---

```{r,echo=FALSE,eval=TRUE}
library(knitr)
args = commandArgs(TRUE)
d = args[1]
rd = read.table(file.path(d, "report.data"), sep="\t")
rd_dic = rd[[2]]
names(rd_dic) = rd[[1]]
```


Report generated: `r format(Sys.time(), "%F %H:%M:%S")` 

Pipeline: `r as.character(rd_dic["version"])` <!-- [(Documentation)](http://wiki.immunbio.mpg.de/wiki/index.php/RNA-seq_pipeline_%28rna-seq-qc.py%29) -->

---

## Data quantities

```{r,echo=FALSE,eval=TRUE}
library(knitr)
args = commandArgs(TRUE)
d = args[1]
d = c("/data/processing/kilpert/test/rna-seq-qc/Ausma/PE_mm10_FULL_hisat_trim/project_report/")
df = read.table(file.path(d, "report.tsv"))
colnames(df)[grepl("_perc$",colnames(df))] = "%"
kable(df, format = "markdown", digits=1)
```

FASTQ: number of sequencing reads, TrimGalore: number of reads after trimming, HISAT/TopHat2: number of mapped reads defined by samtools flags (-f3 -F256) for proper pair + primary alignment, featureCounts/htseq-count: counts of reads mapped to genes, %: percentage compared to the starting number of sequencing reads

---

## Workflow software

`r as.character(rd_dic["bowtie2_ver"])`, 
`r as.character(rd_dic["cutadapt_ver"])`, 
`r as.character(rd_dic["deseq2_ver"])`, 
`r as.character(rd_dic["fastqc_ver"])`, 
`r as.character(rd_dic["feature_counts_ver"])`, 
`r as.character(rd_dic["hisat_ver"])`, 
`r as.character(rd_dic["picardtools_ver"])`, 
`r as.character(rd_dic["R_ver"])`, 
`r as.character(rd_dic["rseqc_ver"])`, 
`r as.character(rd_dic["samtools_ver"])`, 
`r as.character(rd_dic["tophat2_ver"])`, 
`r as.character(rd_dic["trim_galore_ver"])` 

---

## Analysis details

Data type: `r as.character(rd_dic["seq_type"])` 

Data type: `r as.character(rd_dic["library_type"])` 

Genome acronym: `r as.character(rd_dic["genome"])` 

Mapping software: `r as.character(rd_dic["mapping_prg"])`

Counting software: `r as.character(rd_dic["count_prg"])`

Insert metrics by: `r as.character(rd_dic["insert_metrics"])` 

Input directory: `r as.character(rd_dic["main_indir"])` 

Output directory: `r as.character(rd_dic["main_outdir"])` 

FASTA index: `r as.character(rd_dic["fasta_index"])`

Transcriptome index: `r as.character(rd_dic["transcriptome_index"])` 

Genome index: `r as.character(rd_dic["genome_index"])` 

BED: `r as.character(rd_dic["transcriptome_index"])` 

Experiment setup (DESeq2): `r as.character(rd_dic["sample_info"])` 

Pipeline executable: `r as.character(rd_dic["script_path"])`rna-seq-qc.py 

Threads (max. per process): `r as.character(rd_dic["threads"])`

Files processed in parallel: `r as.character(rd_dic["parallel"])`

---

## DESeq2

```{r}

```

---

MPI-IE - Bioinformatics (`r format(Sys.time(), "%Y")`)

