\documentclass[12pt,a4paper]{article}
\usepackage{color}
\usepackage{geometry}
 \geometry{
 a4paper,
%  total={210mm,297mm},
 left=30mm,
 right=20mm,
 top=20mm,
 bottom=20mm,
 }
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{tikz}
% \usepackage{wrapfig}
\usepackage{url}

\title{RNA-seq report}
\author{MPI-IE Bioinformatics}


<<read_report_data, include=FALSE, eval=TRUE, echo=FALSE>>=
library(knitr)
args = commandArgs(TRUE)
d = opts_knit$get("report.dir")
rd = read.table(file.path(d, "Report.data"), sep="\t")
rd_dic = rd[[2]]
names(rd_dic) = rd[[1]]
plots1 = file.path(d, "plots1.png")
plots2 = file.path(d, "plots2.png")
@


\begin{document}

% \maketitle

% \begin{wrapfigure}[0]{r}{0.25\textwidth} %this figure will be at the right
% \includegraphics[width=0.15\textwidth]{\Sexpr{rd_dic["logo"]}}
% \end{wrapfigure}

\large{MPI-IE Bioinformatics}

\Huge{RNA-seq report}\\

\normalsize

\Sexpr{format(Sys.time(), "%F %H:%M:%S")}

% Pipeline: `r as.character(rd_dic["version"])` <!-- [(Documentation)](http://wiki.immunbio.mpg.de/wiki/index.php/RNA-seq_pipeline_%28rna-seq-qc.py%29) -->

\noindent\rule{8cm}{0.4pt}

\noindent

\section{About} This document was automatically generated on finish of the RNA-seq pipeline (\Sexpr{rd_dic["version"]}).


\section{Data quantaties}

<<quantities_table, include=TRUE, eval=TRUE, echo=FALSE>>=
df = read.table(file.path(d, "Report.tsv"))
colnames(df)[grepl("_perc$",colnames(df))] = "%"
kable(df, format="latex", digits=1, booktabs=TRUE, longtable=TRUE)
@
FASTQ: number of sequencing reads, TrimGalore: number of reads after trimming, HISAT2/TopHat2: number of mapped reads defined by samtools flags (SE: -q20 -F256 [min. quality 20, primary alignment]; PE: -f3 -q20 -F256 [properly paired, min. quality 20, primary alignment]), featureCounts/htseq-count: counts of reads mapped to genes, \%: percentage compared to the starting number of sequencing reads


\section{Workflow software}

\Sexpr{rd_dic["bowtie2_ver"]},
\Sexpr{rd_dic["cutadapt_ver"]},
\Sexpr{rd_dic["deseq2_ver"]},
\Sexpr{rd_dic["fastqc_ver"]},
\Sexpr{rd_dic["feature_counts_ver"]},
\Sexpr{rd_dic["hisat_ver"]}, 
\Sexpr{rd_dic["picardtools_ver"]},
\Sexpr{rd_dic["R_ver"]},
\Sexpr{rd_dic["rseqc_ver"]},
\Sexpr{rd_dic["samtools_ver"]},
\Sexpr{rd_dic["tophat2_ver"]},
\Sexpr{rd_dic["trim_galore_ver"]}
\newpage


\IfFileExists{\Sexpr{plots2}}{
\section{Sample correlations}
\begin{figure}[ht]
\centering
\includegraphics[width=1\textwidth,]{\Sexpr{plots2}}
\caption{Correlation between samples based on normalized and rlog-transformed read counts (DESeq2)}
\label{fig:figure1}
\end{figure}
}


\IfFileExists{\Sexpr{plots1}}{
\section{Differential expression}
\begin{figure}[ht]
\centering
\includegraphics[width=1\textwidth,]{\Sexpr{plots1}}
\caption{Differentially expressed genes based on normalized counts (DESeq2)}
\label{fig:figure2}
\end{figure}

\newpage
}


\section{Analysis details}

\begin{description}

\item[Data type:] \path{\Sexpr{rd_dic["seq_type"]}}

\item[Library type:] \path{\Sexpr{rd_dic["library_type"]}}

\item[Genome acronym:] \path{\Sexpr{rd_dic["genome"]}}

\item[Mapping software:] \path{\Sexpr{rd_dic["mapping_prg"]}}

\item[Counting software:] \path{\Sexpr{rd_dic["count_prg"]}}

\item[Insert metrics:] \path{\Sexpr{rd_dic["insert_metrics"]}}

\item[Input directory:] \path{\Sexpr{rd_dic["main_indir"]}}

\item[Output directory:] \path{\Sexpr{rd_dic["main_outdir"]}}

\item[FASTA index:] \path{\Sexpr{rd_dic["fasta_index"]}}

\item[Transcriptome index:] \path{\Sexpr{rd_dic["transcriptome_index"]}}

\item[Genome index:] \path{\Sexpr{rd_dic["genome_index"]}}

\item[BED:] \path{\Sexpr{rd_dic["bed"]}}

\item[Experiment setup (DESeq2):] \path{\Sexpr{rd_dic["sample_info"]}}

\item[Pipeline directory:] \path{\Sexpr{rd_dic["script_path"]}}

\item[Threads (max. per process):] \path{\Sexpr{rd_dic["threads"]}}

\item[Files processed in parallel:] \path{\Sexpr{rd_dic["parallel"]}}

\end{description}


\end{document}
